src/main/python/systemds/script_building/dag.py

```diff
@@ -45,9 +45,10 @@ class DAGNode(ABC):
     sds_context: 'SystemDSContext'
     _unnamed_input_nodes: Sequence[Union['DAGNode', str, int, float, bool]]
     _named_input_nodes: Dict[str, Union['DAGNode', str, int, float, bool]]
-    _output_type: OutputType
+    _outputs: List[Tuple[str, OutputType]]
+    _output_nodes: Dict[str, Union['DAGNode']]
     _is_python_local_data: bool
-    _number_of_outputs: int
+    _dml_name: str
```

Comment:
I can create this structure if you want, but in my opinion a single universal class that can handle any cases has less complexity and is thus easier to maintain. The variable _dml_name is needed so I can perform DFS on the DAG without evaluating paths more than once in case more than one output of a node are used as input in the DAG